File: frontend/types/request.ts
export interface KeyValueItem {
  key: string;
  value: string;
  description?: string;
  enabled: boolean;
}

export interface AuthData {
  token?: string;
  username?: string;
  password?: string;
}

export type HttpMethod = 'GET' | 'POST' | 'PUT' | 'PATCH' | 'DELETE' | 'HEAD' | 'OPTIONS';
export type BodyType = 'none' | 'raw' | 'json' | 'text' | 'xml' | 'form-data' | 'x-www-form-urlencoded';
export type AuthType = 'none' | 'bearer' | 'basic';

export interface RequestItem {
  id: string;
  collection_id: string;
  folder_id: string | null;
  name: string;
  method: HttpMethod;
  url: string;
  params: KeyValueItem[];
  headers: KeyValueItem[];
  body_type: BodyType;
  body_content: string | null;
  auth_type: AuthType;
  auth_data: AuthData;
  description: string | null;
  is_favorite: boolean;
  created_at: string;
  updated_at: string;
}

export interface RequestCreate {
  collection_id: string;
  folder_id?: string | null;
  name: string;
  method: HttpMethod;
  url: string;
  params?: KeyValueItem[];
  headers?: KeyValueItem[];
  body_type?: BodyType;
  body_content?: string | null;
  auth_type?: AuthType;
  auth_data?: AuthData;
  description?: string | null;
}

export interface RequestUpdate {
  folder_id?: string | null;
  name?: string;
  method?: HttpMethod;
  url?: string;
  params?: KeyValueItem[];
  headers?: KeyValueItem[];
  body_type?: BodyType;
  body_content?: string | null;
  auth_type?: AuthType;
  auth_data?: AuthData;
  description?: string | null;
  is_favorite?: boolean;
}

File: backend/schemas/request.py
from datetime import datetime
from typing import Optional, Any
from pydantic import BaseModel, Field


class KeyValueItem(BaseModel):
    """Reusable key-value pair with optional description and enabled flag."""
    key: str = ""
    value: str = ""
    description: Optional[str] = None
    enabled: bool = True


class AuthData(BaseModel):
    token: Optional[str] = None
    username: Optional[str] = None
    password: Optional[str] = None


class RequestCreate(BaseModel):
    collection_id: str
    folder_id: Optional[str] = None
    name: str = Field(..., min_length=1, max_length=255)
    method: str = Field("GET", pattern="^(GET|POST|PUT|PATCH|DELETE|HEAD|OPTIONS)$")
    url: str = ""
    params: list[KeyValueItem] = []
    headers: list[KeyValueItem] = []
    body_type: str = "none"
    body_content: Optional[str] = None
    auth_type: str = "none"
    auth_data: dict[str, Any] = {}
    description: Optional[str] = None
    is_favorite: bool = False


class RequestUpdate(BaseModel):
    folder_id: Optional[str] = None
    name: Optional[str] = Field(None, min_length=1, max_length=255)
    method: Optional[str] = Field(None, pattern="^(GET|POST|PUT|PATCH|DELETE|HEAD|OPTIONS)$")
    url: Optional[str] = None
    params: Optional[list[KeyValueItem]] = None
    headers: Optional[list[KeyValueItem]] = None
    body_type: Optional[str] = None
    body_content: Optional[str] = None
    auth_type: Optional[str] = None
    auth_data: Optional[dict[str, Any]] = None
    description: Optional[str] = None
    is_favorite: Optional[bool] = None


class RequestResponse(BaseModel):
    id: str
    collection_id: str
    folder_id: Optional[str]
    name: str
    method: str
    url: str
    params: list[Any]
    headers: list[Any]
    body_type: str
    body_content: Optional[str]
    auth_type: str
    auth_data: dict[str, Any]
    description: Optional[str]
    is_favorite: bool
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class RequestMoveRequest(BaseModel):
    collection_id: str
    folder_id: Optional[str] = None
